[miopen-hip/opencl] Build failure boost inline

When I try to build (using the aur PKGBUILD, nothing changed), I get error messages originating from boost headers, centering around the keyword `noinline`. See the build log here: https://gist.github.com/bytbox/019ab8439b83cf3edd7965c2265187a5

The closest thing I can find online is [this](https://github.com/BVLC/caffe/issues/171). That's old enough that I kind of doubt it's related.

No -testing repositories are enabled for me. This is my first attempt to build, so I have no idea if this is a new problem (and also it's likely just me doing something wrong).

Hardware (surely not relevant): Navi 10 [Radeon RX 5600 OEM/5600 XT / 5700/5700 XT]

And I have the following error when compiling.
```
[3647/4636] cd /var/tmp/pamac-build-tak/python-pytorch-rocm/src/pytorch-1.7.1-rocm/build/caffe2/CMakeFiles/torch_hip.dir/__/aten/src/THH && /usr/bin/cmake -E make_directory /var/tmp/pamac-build-tak/python-pytorch-rocm/src/pytorch-1.7.1-rocm/build/caffe2/CMakeFiles/torch_hip.dir/__/aten/src/THH/. && /usr/bin/cmake -D verbose:BOOL=OFF -D build_configuration:STRING=RELEASE -D generated_file:STRING=/var/tmp/pamac-build-tak/python-pytorch-rocm/src/pytorch-1.7.1-rocm/build/caffe2/CMakeFiles/torch_hip.dir/__/aten/src/THH/./torch_hip_generated_THHTensorIndex.hip.o -P /var/tmp/pamac-build-tak/python-pytorch-rocm/src/pytorch-1.7.1-rocm/build/caffe2/CMakeFiles/torch_hip.dir/__/aten/src/THH/torch_hip_generated_THHTensorIndex.hip.o.cmake
ninja: build stopped: subcommand failed.
Building wheel torch-1.7.1
-- Building version 1.7.1
cmake -GNinja -DBUILD_BINARY=ON -DBUILD_CUSTOM_PROTOBUF=OFF -DBUILD_PYTHON=True -DBUILD_TEST=True -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=/var/tmp/pamac-build-tak/python-pytorch-rocm/src/pytorch-1.7.1-rocm/torch -DCMAKE_PREFIX_PATH=/usr/lib/python3.8/site-packages -DCUDNN_INCLUDE_DIR=/usr/include -DCUDNN_LIBRARY=/usr/lib -DDISABLE_AVX2=1 -DDISABLE_AVX512F=1 -DDISABLE_FMA4=1 -DDISABLE_SSE4=1 -DNUMPY_INCLUDE_DIR=/usr/lib/python3.8/site-packages/numpy/core/include -DPYTHON_EXECUTABLE=/usr/bin/python -DPYTHON_INCLUDE_DIR=/usr/include/python3.8 -DPYTHON_LIBRARY=/usr/lib/libpython3.8.so.1.0 -DTORCH_BUILD_VERSION=1.7.1 -DUSE_CUDA=OFF -DUSE_FFMPEG=ON -DUSE_GFLAGS=ON -DUSE_GLOG=ON -DUSE_MKLDNN=ON -DUSE_NUMPY=True -DUSE_ROCM=ON -DUSE_SYSTEM_NCCL=ON /var/tmp/pamac-build-tak/python-pytorch-rocm/src/pytorch-1.7.1-rocm
cmake --build . --target install --config Release -- -j 16
Traceback (most recent call last):
  File "setup.py", line 760, in <module>
    build_deps()
  File "setup.py", line 310, in build_deps
    build_caffe2(version=version,
  File "/var/tmp/pamac-build-tak/python-pytorch-rocm/src/pytorch-1.7.1-rocm/tools/build_pytorch_libs.py", line 62, in build_caffe2
    cmake.build(my_env)
  File "/var/tmp/pamac-build-tak/python-pytorch-rocm/src/pytorch-1.7.1-rocm/tools/setup_helpers/cmake.py", line 349, in build
    self.run(build_args, my_env)
  File "/var/tmp/pamac-build-tak/python-pytorch-rocm/src/pytorch-1.7.1-rocm/tools/setup_helpers/cmake.py", line 141, in run
    check_call(command, cwd=self.build_dir, env=env)
  File "/usr/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['cmake', '--build', '.', '--target', 'install', '--config', 'Release', '--', '-j', '16']' returned non-zero exit status 1.
==> ERROR: A failure occurred in build().
    Aborting...
```

I have been having the same issue and found that the issue can be avoided by enabling the `BOOST_CONTAINER_DISABLE_NOINLINE` flag. This flag was added in boostorg/container@c08ead7f4b4e4aa5674151e180f05fcc0a05e644 but is otherwise entirely undocumented.

Although it removes the main issue, the linking of `libMIOpen.so` fails later due ot the following error:

```
/sbin/ld: CMakeFiles/MIOpen.dir/mdg_expr.cpp.o: in function `boost::spirit::utree boost::spirit::function_impl_invert::eval<bool>(bool const&)':
mdg_expr.cpp:(.text+0x0): multiple definition of `boost::spirit::utree boost::spirit::function_impl_invert::eval<bool>(bool const&)'; CMakeFiles/MIOpen.dir/md_graph.cpp.o:md_graph.cpp:(.text+0x0): first defined here
clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [src/CMakeFiles/MIOpen.dir/build.make:2662: lib/libMIOpen.so.1.0] Error 1
make[2]: Leaving directory '/tmp/makepkg/miopen-hip/src/build'
make[1]: *** [CMakeFiles/Makefile2:2636: src/CMakeFiles/MIOpen.dir/all] Error 2
make[1]: Leaving directory '/tmp/makepkg/miopen-hip/src/build'
make: *** [Makefile:182: all] Error 2
make: Leaving directory '/tmp/makepkg/miopen-hip/src/build'
```

I don't know whether this later error is related to the addition of the flag yet.

--- a/miopen-hip/.SRCINFO
+++ b/miopen-hip/.SRCINFO
@@ -1,7 +1,7 @@
 pkgbase = miopen-hip
 	pkgdesc = AMD's Machine Intelligence Library (HIP backend)
 	pkgver = 4.0.0
-	pkgrel = 1
+	pkgrel = 2
 	url = https://github.com/ROCmSoftwarePlatform/MIOpen
 	arch = x86_64
 	license = custom:NCSAOSL
@@ -22,3 +22,4 @@ pkgbase = miopen-hip
 	sha256sums = 84c6c17be9c1a9cd0d3a2af283433f64b07a4b9941349f498e40fed82fb205a6
 
 pkgname = miopen-hip
+

--- a/miopen-hip/PKGBUILD
+++ b/miopen-hip/PKGBUILD
@@ -1,8 +1,9 @@
 # Maintainer: acxz <akashpatel at yahoo dot com>
+# Contributor: JP-Ellis <josh@jpellis.me>
 
 pkgname=miopen-hip
 pkgver=4.0.0
-pkgrel=1
+pkgrel=2
 pkgdesc="AMD's Machine Intelligence Library (HIP backend)"
 arch=('x86_64')
 url="https://github.com/ROCmSoftwarePlatform/MIOpen"
@@ -15,16 +16,13 @@ source=("$pkgname-$pkgver.tar.gz::$url/archive/rocm-$pkgver.tar.gz")
 sha256sums=('84c6c17be9c1a9cd0d3a2af283433f64b07a4b9941349f498e40fed82fb205a6')
 
 build() {
-  CXXFLAGS="$CXXFLAGS -DHALF_ENABLE_F16C_INTRINSICS=0 -isystem /opt/rocm/llvm/lib/clang/11.0.0" \
-  CPPFLAGS="$CPPFLAGS -DHALF_ENABLE_F16C_INTRINSICS=0 -isystem /opt/rocm/llvm/lib/clang/11.0.0" \
-  CXX=/opt/rocm/hip/bin/hipcc \
+  CXX=/opt/rocm/llvm/bin/clang++ \
+  CXXFLAGS="$CXXFLAGS -DBOOST_CONTAINER_DISABLE_NOINLINE=ON" \
   cmake -B build -Wno-dev \
         -S "MIOpen-rocm-$pkgver" \
-        -DCMAKE_INSTALL_PREFIX=/opt/rocm \
         -DMIOPEN_BACKEND=HIP \
-        -DMIOPEN_HIP_COMPILER=/opt/rocm/llvm/bin/clang++ \
-        -DHALF_INCLUDE_DIR=/usr/include/half \
-        -DBoost_NO_BOOST_CMAKE=ON
+        -DCMAKE_INSTALL_PREFIX=/opt/rocm \
+        -DHALF_INCLUDE_DIR=/usr/include/half
 
   make -C build
 }

